pkg/controllers/nodeclass/validation.go

```diff
@@ -324,19 +339,98 @@ func mockLaunchTemplateConfig() []ec2types.FleetLaunchTemplateConfigRequest {
-		},
-		{
-			Name:         "m6g.large",
-			Requirements: scheduling.NewRequirements(scheduling.NewRequirement(corev1.LabelArchStable, corev1.NodeSelectorOpIn, karpv1.ArchitectureArm64)),
+
+	// Select an instance type to use for validation. If NodePools exist for this NodeClass, we'll use an instance type
+	// selected by one of those NodePools. We should also prioritize an InstanceType which will launch with a non-GPU AMI,
+	// since GPU AMIs may have a larger snapshot size than that supported by the NodeClass' blockDeviceMappings.
+	// Issue: https://github.com/aws/karpenter-provider-aws/issues/7928
+	amiMap := amifamily.MapToInstanceTypes(v.getInstanceTypesForNodeClass(ctx, nodeClass), nodeClass.Status.AMIs)
+	var instanceTypes []*cloudprovider.InstanceType
+	for _, ami := range nodeClass.Status.AMIs {
+		if len(amiMap[ami.ID]) == 0 {
+			continue
+		}
+		amiRequirements := scheduling.NewNodeSelectorRequirements(ami.Requirements...)
+		if amiRequirements.IsCompatible(amifamily.VariantStandard.Requirements()) {
+			instanceTypes = append(instanceTypes, amiMap[ami.ID]...)
+		}
+	}
+	if len(instanceTypes) == 0 && len(amiMap) != 0 {
+		instanceTypes = lo.Flatten(lo.Values(amiMap))
+	}
+	// If there weren't any matching instance types, we should fallback to some defaults. There's an instance type included
+	// for both x86_64 and arm64 architectures, ensuring that there will be a matching AMI.
+	if len(instanceTypes) == 0 {
+		instanceTypes = []*cloudprovider.InstanceType{
+			{
+				Name: string(ec2types.InstanceTypeM5Large),
+				Requirements: scheduling.NewRequirements(append(
+					lo.Values(amifamily.VariantStandard.Requirements()),
+					scheduling.NewRequirement(corev1.LabelArchStable, corev1.NodeSelectorOpIn, karpv1.ArchitectureAmd64),
+				)...),
+			},
+			{
+				Name: string(ec2types.InstanceTypeM6gLarge),
+				Requirements: scheduling.NewRequirements(append(
+					lo.Values(amifamily.VariantStandard.Requirements()),
+					scheduling.NewRequirement(corev1.LabelArchStable, corev1.NodeSelectorOpIn, karpv1.ArchitectureArm64),
+				)...),
+			},
+		}
+	}
+	opts, err := v.amiResolver.Resolve(nodeClass, nodeClaim, instanceTypes, karpv1.CapacityTypeOnDemand, amiOptions)
+	if err != nil {
+		return nil, err
+	}
+	return opts[0], nil
+}
+
+// getInstanceTypesForNodeClass returns the set of instances which could be launched using this NodeClass based on the
+// requirements of linked NodePools. If no NodePools exist for the given NodeClass, this function returns two default
+// instance types (one x86_64 and one arm64).
+func (v *Validation) getInstanceTypesForNodeClass(ctx context.Context, nodeClass *v1.EC2NodeClass) []*cloudprovider.InstanceType {
+	instanceTypes, err := v.cloudProvider.GetInstanceTypes(ctx, &karpv1.NodePool{Spec: karpv1.NodePoolSpec{Template: karpv1.NodeClaimTemplate{Spec: karpv1.NodeClaimTemplateSpec{
```

Comment:
I get that you would have to call GetInstanceTypes below multiple times, but it's relatively cheap (since they are cached) and would probably prevent bugs in the future just in case there was something that was injected when you specified a particular NodePool inside of the GetInstanceTypes() call

Reply:
I actually needed to switch to calling the instance type provider here directly, so it ends up remaining largely the same. I needed to make this switch because otherwise you always saw failures on the initial reconciliation. Since the cloudprovider method gets the NodeClass from the API server, and we haven't applied the updated status yet, it will always fail to describe instance types due to missing subnets.